System.Messaging/System/Messaging/Interop/SafeHandles.cs

```diff
@@ -0,0 +1,143 @@
+            SafeNativeMethods.MQCloseQueue(this.handle);
+
+            return true;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return base.IsInvalid || IsClosed; }
+        }
+
+
+        // A subclass needed to express InvalidHandle. The reason is that CLR notices that 
+        // ReleaseHandle requires a call to MQRT.DLL, and throws in the ctor if MQRT.DLL is not available,
+        // even though CTOR ITSELF DOES NOT REQUIRE MQRT.DLL. 
+        // We address this by defining a NOOP ReleaseHandle
+        sealed class InvalidMessageQueueHandle : MessageQueueHandle
+        {
+
+            protected override bool ReleaseHandle()
+            {
+                return true;
+            }
+        }
+    }
+
+
+    internal class CursorHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+
+        public static readonly CursorHandle NullHandle = new InvalidCursorHandle();
+
+        protected CursorHandle() : base(true) { }
+
+        protected override bool ReleaseHandle()
+        {
+            SafeNativeMethods.MQCloseCursor(this.handle);
+
+            return true;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return base.IsInvalid || IsClosed; }
+        }
+
+
+        // A subclass needed to express InvalidHandle. The reason is that CLR notices that 
+        // ReleaseHandle requires a call to MQRT.DLL, and throws in the ctor if MQRT.DLL is not available,
+        // even though CTOR ITSELF DOES NOT REQUIRE MQRT.DLL. 
+        // We address this by defining a NOOP ReleaseHandle
+        sealed class InvalidCursorHandle : CursorHandle
+        {
+
+            protected override bool ReleaseHandle()
+            {
+                return true;
+            }
+        }
+    }
+
+
+    internal class LocatorHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+
+        public static readonly LocatorHandle InvalidHandle = new InvalidLocatorHandle();
+
+        protected LocatorHandle() : base(true) { }
+
+        protected override bool ReleaseHandle()
+        {
+            SafeNativeMethods.MQLocateEnd(this.handle);
+
+            return true;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return base.IsInvalid || IsClosed; }
+        }
+
+
+        // A subclass needed to express InvalidHandle. The reason is that CLR notices that 
+        // ReleaseHandle requires a call to MQRT.DLL, and throws in the ctor if MQRT.DLL is not available,
+        // even though CTOR ITSELF DOES NOT REQUIRE MQRT.DLL. 
+        // We address this by defining a NOOP ReleaseHandle
+        sealed class InvalidLocatorHandle : LocatorHandle
+        {
+
+            protected override bool ReleaseHandle()
+            {
+                return true;
+            }
+        }
+    }
+
+
+
+
+    internal sealed class SecurityContextHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+
+        internal SecurityContextHandle(IntPtr existingHandle)
```

Comment:
A note to anyone intending to reuse this code.

System.Messaging.MessageQueue.GetSecurityContext() always throws MissingMethodException because SecurityContextHandle lacks a parameterless (perhaps private) constructor that Platform Invoke would call here: <https://github.com/microsoft/referencesource/blob/08847a396ce6da2253bedece91334034d8b98d32/System.Messaging/System/Messaging/Interop/NativeMethods.cs#L320-L334>

I don't expect the bug to be fixed in .NET Framework (reported to connect.microsoft.com in 2010) or in Reference Source, but perhaps it can be fixed in the .NET Core port.